Implement fee-reserve

prusnak opened the pull request:
My shot at fixing/implementing https://github.com/getAlby/lndhub.go/issues/183

It's rather naive and I am not sure if having the check at one place only is enough.

Someone with more Go expertise than me should carefully review.

And yeah, this PR does not contain any tests ... :-/

prusnak commented on README.md:
```diff
@@ -50,6 +50,7 @@ vim .env # edit your config
 + `ENABLE_PROMETHEUS`: (default: false) Enable Prometheus metrics to be exposed
 + `PROMETHEUS_PORT`: (default: 9092) Prometheus port (path: `/metrics`)
 + `WEBHOOK_URL`: Optional. Callback URL for incoming and outgoing payment events, see below.
++ `FEE_RESERVE`: (default: 0) Keep fee reserve for each user (in satoshi)
```
Are balances kept in satoshi or milli-satoshi?

kiwiidb reviewed (commented):
Thanks for your PR! The fee-reserve topic is not easy and it's something that we have debated internally for quite a bit already. Some remarks about this PR:

- users with a balance below `Config.FeeReserve` would be unable to send any payments. 
-  it would still be possible for users to go below 0 because the fee of a large enough payment will be larger than any fixed reserve.
- (Unrelated) the balance check actually already happens [here](https://github.com/getAlby/lndhub.go/blob/main/controllers/payinvoice.ctrl.go#L98) and [here](https://github.com/getAlby/lndhub.go/blob/main/controllers/keysend.ctrl.go#L88) (code duplication I know), so it's better to do the check there.

I agree that we can make it configurable for operators if they want a fee reserve or not. But if they want a fee reserve, this should just be a boolean option and the fee reserve should be exactly the same as the fee limit, which is calculated [here](https://github.com/getAlby/lndhub.go/blob/main/lib/service/invoices.go#L179). Adding more numerical limits makes things too complex imo.

prusnak commented:
> The fee-reserve topic is not easy and it's something that we have debated internally for quite a bit already.

Yeah, I totally get that this is a complex topic.

I addressed your comments and force-pushed 361ae94

kiwiidb reviewed (changes requested):
There are some static check failures because of the now-undefined `calcFeeLimit(invoice)` function in `invoices_test.go`, and .
Other than that LGTM, since this is opt-in, I could implement tests in a seperate PR.

kiwiidb reviewed (approved):
I approve. @bumi gonna leave this open for a couple of days so you can have a look.

bumi reviewed (approved):
yeah, nice! I think that's a great and simple addition for this topic! thanks!

a test would be nice? sending 100% of a user's balance with fee-reserve enabled should fail.

kiwiidb commented:
Tests are currently failing because they run against our regtest LND nodes, which seem to have some issues the last couple of days. I'm looking into replacing the LND connection with an in-memory mock for the integration tests.

prusnak commented:
Rebased and force-pushed
